### PR TITLE
feature: writer

### PR DESCRIPTION
**Pull request by josephcopenhaver**
Opening so unit tests can get interactive feedback from rabbit ai.

This PR if it is merged will add a writer that is configured to take out as much guesswork as possible and ensure the document is a proper utf8 document.

## Summary by CodeRabbit

- **New Features**
  - Introduced an advanced CSV writer that enables flexible CSV generation with customizable headers, comments, and formatting options.
  - Enhanced handling of special characters, including robust escaping and quotation, ensuring accurate data representation.
  
- **Tests**
  - Extended test coverage to validate error handling and ensure consistent behavior across various edge case scenarios.

### CODE REVIEW

File: writer.go [hunk 1]
```diff
@@ -0,0 +1,925 @@
...
+//
+// Should any configuration options require post-flight
+// checks they will be implemented here.
+//
+// It will never attempt to close the underlying writer
+// instance.
+func (w *Writer) Close() error {
+	w.setErr(ErrWriterClosed)
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply writer.go:336]: The last line in the function docstring states that the underlying writer is never closed. I'll add a bit more to make it fool proof that the calling context must handle this detail.

File: writer.go [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+	recordSep      [2]rune
+	numFields      int
+	fieldSeparator rune
+	quote          rune
+	escape         rune
+	numFieldsSet   bool
+	errOnNonUTF8   bool
+	escapeSet      bool
+	recordSepLen   int8
+}
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply writer.go:32]: It's all read-only and has been audited, as the comments have mentioned.

File: writer.go [hunk 3]
```diff
@@ -0,0 +1,846 @@
...
+		errOnNonUTF8:   true,
+	}
+
+	for _, f := range options {
+		f(&cfg)
+	}
+
+	if err := cfg.validate(); err != nil {
+		return nil, errors.Join(ErrBadConfig, err)
+	}
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply writer.go:254]: It's defined in another file, you just do not see it. Check writer.go.

File: writer.go [hunk 4]
```diff
@@ -0,0 +1,846 @@
...
+}
+
+func (cfg *wCfg) validate() error {
+	if cfg.writer == nil {
+		return errors.New("writer must not be nil")
+	}
+
+	if cfg.recordSepLen == 0 {
+		return ErrBadRecordSeparator
+	}
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply writer.go:149]: No. it's a config time error and I'm not going to offer this low-level detail to the calling context to do something intelligent with that specific context.

They should instead trickle that up to a human to ingest as something has been misconfigured. The human readable message will be descriptive enough.

File: writer.go [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+		return errors.New("invalid field separator value")
+	}
+
+	if !validUtf8Rune(cfg.quote) || isNewlineRuneForWrite(cfg.fieldSeparator) {
+		return errors.New("invalid quote value")
+	}
+
+	if cfg.escapeSet && (!validUtf8Rune(cfg.escape) || isNewlineRuneForWrite(cfg.fieldSeparator)) {
+		return errors.New("invalid escape value")
+	}
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply writer.go:170]: This will be resolved in the next force commit.

File: writer.go [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+	trimHeaders     bool
+	headersSet      bool
+	commentRuneSet  bool
+	commentLinesSet bool
+	includeBOM      bool
+}
+
+type WriteHeaderOption func(*whCfg)
+
+type WriteHeaderOptions struct{}
```
Comment (josephcopenhaver) [inline writer.go:352]: Write a docstrings about proper initialization

File: writer.go [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+	var escapedEscape [utf8.UTFMax * 2]byte
+	var recordSepBytes [utf8.UTFMax]byte
+	var escapedQuoteByteLen, escapedEscapeByteLen, recordSepByteLen, twoQuotesByteLen int8
+	if cfg.escapeSet {
+		n := utf8.EncodeRune(escapedQuote[:], cfg.escape)
+		n += utf8.EncodeRune(escapedQuote[n:], cfg.quote)
+		escapedQuoteByteLen = int8(n)
+
+		n = utf8.EncodeRune(escapedEscape[:], cfg.escape)
+		n += utf8.EncodeRune(escapedEscape[n:], cfg.escape)
```
Comment (josephcopenhaver) [inline writer.go:267]: Just slice copy instead of falling Encode twice for the same rune.

File: functional_reader_test.go [hunk 8]
```diff
@@ -65,7 +65,7 @@ func (tc *functionalReaderTestCase) Run(t *testing.T) {
 				}
 
 				v, err := csv.NewReader(opts...)
-				if tc.hasNewReaderErr || len(tc.newReaderErrIs) > 0 || len(tc.newReaderErrAs) > 0 || tc.newReaderErrStr != "" {
+				if tc.hasNewReaderErr || len(tc.newReaderErrIs) > 0 || len(tc.newReaderErrAs) > 0 || len(tc.newReaderErrAs) > 0 || len(tc.newReaderErrAsNot) > 0 || tc.newReaderErrStr != "" {
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply functional_reader_test.go]: yup, resolved in next commit

File: internal/cmd/generate/process_field.go.tmpl [hunk 9]
```diff
@@ -0,0 +1,55 @@
+func (w *Writer) processField_escape{{if .EscapeSet}}S{{else}}Uns{{end}}et_quote{{if .QuoteForced}}F{{else}}Unf{{end}}orced(v []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply internal/cmd/generate/process_field.go.tmpl:1]: No thank you. As commented elsewhere, -1 means that no additional side-channel buffering has occurred and the raw record can be read transparently in a higher calling context should there be no error countered traversing the input field value for csv encoding.

File: gen_strategies.go [hunk 10]
```diff
@@ -795,3 +795,207 @@ func (r *Reader) prepareRow_memclearEnabled() bool {
...
+
+		break
+	}
+
+	si2, err := w.escapeCharsWithEscapeCheck(v[si:], i-si)
+	if err != nil {
+		return -1, err
+	}
+
+	return si + si2, nil
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply gen_strategies.go:890]: As mentioned elsewhere, this is generated code in a hot path that has only what is required for that path. refactoring for ease of maintaining is not a priority because all variations are already maintained in one template file.

File: gen_strategies.go [hunk 11]
```diff
@@ -795,3 +795,207 @@ func (r *Reader) prepareRow_memclearEnabled() bool {
...
+
+		break
+	}
+
+	si2, err := w.escapeChars(v[si:], i-si)
+	if err != nil {
+		return -1, err
+	}
+
+	return si + si2, nil
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply gen_strategies.go:840]: No thank you. As commented elsewhere, -1 means that no additional side-channel buffering has occurred and the raw record can be read transparently in a higher calling context should there be no error encountered traversing the input field value for csv encoding.

File: writer.go [hunk 12]
```diff
@@ -0,0 +1,810 @@
...
+	recordSepLen            int8
+	twoQuotesByteLen        int8
+	escapedQuoteByteLen     int8
+	escapedEscapeByteLen    int8
+	recordSepByteLen        int8
+	escapeSet               bool
+	errOnNonUTF8            bool
+	headerWritten           bool
+	recordWritten           bool
+}
```
[the remark this answers is not in the document]
Comment (josephcopenhaver) [reply writer.go:257]: Maybe in the future. The exported NewWriter function has this note already.
